res/res_pjsip_device_features.c

```diff
@@ -0,0 +1,1005 @@
+		<synopsis>
+			Set or get a device feature of a PJSIP endpoint.
+		</synopsis>
+		<syntax>
+			<parameter name="endpoint" required="true">
+				<para>Name of PJSIP endpoint, without the technology.</para>
+			</parameter>
+			<parameter name="feature" required="true">
+				<para>Name of the feature.</para>
+				<para>Must be one of:</para>
+				<enumlist>
+					<enum name="donotdisturb"/>
+					<enum name="callforwardingalways"/>
+					<enum name="callforwardingbusy"/>
+					<enum name="callforwardingnoanswer"/>
+					<enum name="ringcount">
+						<para>This applies only to reading the function.</para>
+						<para>To set the ring count, pass it as the second data value with callforwardingnoanswer.</para>
+					</enum>
+				</enumlist>
+			</parameter>
+		</syntax>
+		<description>
+			<para>Can be used to provide device feature key synchronization with supporting IP phones.</para>
+			<para>This module does not actually provide any features. It is used to synchronize the indication
+			of features on an IP phone when the server itself provides features whose indications are synchronized
+			by this module.</para>
+			<para>For example:</para>
+			<example title="Enable Do Not Disturb indication for endpoint Phone1">
+			same => n,Set(PJSIP_DEVICE_FEATURES(Phone1,donotdisturb)=enabled)
+			</example>
+			<example title="Set Call Forwarding No Answer to 5551212 after 5 rings">
+			same => n,Set(PJSIP_DEVICE_FEATURES(Phone1,callforwardingnoanswer)=5551212,5)
+			</example>
+			<example title="Disable Call Forwarding Busy">
+			same => n,Set(PJSIP_DEVICE_FEATURES(Phone1,callforwardingbusy)=)
+			</example>
+			<example title="Retrieve current Do Not Disturb status">
+			same => n,NoOp(${PJSIP_DEVICE_FEATURES(Phone1,donotdisturb)})
+			</example>
+			<example title="Retrieve current number of rings">
+			same => n,NoOp(${PJSIP_DEVICE_FEATURES(Phone1,ringcount)})
+			</example>
+		</description>
+	</function>
+***/
+
+/*!
+ * \brief A subscription for synchronized device feature state
+ *
+ * This structure acts as the owner for the underlying SIP subscription.
+ */
+struct feature_state_subscription {
+	/*! The SIP subscription */
+	struct ast_sip_subscription *sip_sub;
+	/*! The serializer to use for notifications */
+	struct ast_taskprocessor *serializer;
+	/*! Info used by res_pjsip_features_body_generator */
+	struct ast_sip_device_feature_sync_data sync_data;
+	/* Whether we have data about this feature. */
+	unsigned int have_dnd:1;
+	unsigned int have_callforwardalways:1;
+	unsigned int have_callforwardbusy:1;
+	unsigned int have_callforwardnoanswer:1;
+};
+
+/* Whether we should auto-approve device feature changes.
+ * If disabled, system will use the AMI event to update feature data and write PJSIP_DEVICE_FEATURES to trigger a NOTIFY.
+ * If enabled, we will "auto approve": internal cache will automatically update per the request and a NOTIFY will be sent. */
+static int auto_approve = 0;
```

Comment:
It's constant. I could not find the code which can change this.
I think this should be PJSIP Endpoint parameter, it means some endpoints are allowed to change it automatically, some endpoints not.

Reply:
This absolutely needs to be configurable per-endpoint.

Reply:
The endpoint property. For example

ast_sorcery_object_field_register(sip_sorcery, "endpoint", "allow_device_features_sync", "no", OPT_BOOL_T, 1, FLDSET(struct ast_sip_endpoint, allow_device_features_sync));